controllers/backup_controller_test.go

```diff
@@ -0,0 +1,273 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	cassdcapi "github.com/k8ssandra/cass-operator/operator/pkg/apis/cassandra/v1beta1"
+	api "github.com/k8ssandra/medusa-operator/api/v1alpha1"
+	"github.com/k8ssandra/medusa-operator/pkg/medusa"
+	"github.com/k8ssandra/medusa-operator/pkg/pb"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"reflect"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strconv"
+	"sync"
+	"testing"
+)
+
+func testBackupDatacenter(t *testing.T, ctx context.Context, namespace string) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	backupName := "test-backup"
+
+	dcKey := types.NamespacedName{
+		Name:      TestCassandraDatacenterName,
+		Namespace: namespace,
+	}
+
+	dc := &cassdcapi.CassandraDatacenter{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        dcKey.Name,
+			Namespace:   dcKey.Namespace,
+			Annotations: map[string]string{},
+		},
+		Spec: cassdcapi.CassandraDatacenterSpec{
+			ClusterName:   "test-dc",
+			ServerType:    "cassandra",
+			ServerVersion: "3.11.7",
+			Size:          3,
+			StorageConfig: cassdcapi.StorageConfig{
+				CassandraDataVolumeClaimSpec: &corev1.PersistentVolumeClaimSpec{
+					VolumeName: "data",
+				},
+			},
+			Racks: []cassdcapi.Rack{
+				{
+					Name: "rack1",
+				},
+			},
+			PodTemplateSpec: &corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{},
+					InitContainers: []corev1.Container{
+						{
+							Name: "medusa-restore",
+							Env: []corev1.EnvVar{
+								{
+									Name:  "MEDUSA_MODE",
+									Value: "RESTORE",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	t.Logf("creating datacenter %s", dcKey)
+
+	err := testClient.Create(ctx, dc)
+	assert.NoError(err, "failed to create datacenter")
+
+	t.Logf("creating datacenter service")
+
+	dcServiceKey := types.NamespacedName{Namespace: dcKey.Namespace, Name: dc.GetAllPodsServiceName()}
+	dcService := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: dcServiceKey.Namespace,
+			Name:      dcServiceKey.Name,
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{
+				cassdcapi.ClusterLabel: dc.Spec.ClusterName,
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Name: "cql",
+					Port: 9042,
+				},
+			},
+		},
+	}
+	err = testClient.Create(ctx, dcService)
+	assert.NoError(err, "failed to create datacenter service %s", dcServiceKey)
+
+	t.Log("creating datacenter pods")
+	createDatacenterPods(t, ctx, dc)
+
+	t.Log("make the datacenter ready")
+	patch := client.MergeFrom(dc.DeepCopy())
```

Comment:
Not really visible in this PR, but is this MergeFrom / Patch call a strict one with the testClient? Thus following Update's semantics and can fail if we're trying to update old one? Also, no need for waiting for these events to happen (will the client block)?

Reply:
>Not really visible in this PR, but is this MergeFrom / Patch call a strict one with the testClient? Thus following Update's semantics and can fail if we're trying to update old one?

I don't think there is any issue with this `MergeFrom`. With that said, I think using `client.MergeFromWithOptimisticLock{}` is probably a safer default so I will make the changes accordingly.

>Also, no need for waiting for these events to happen (will the client block)?

That is correct. Waiting needs to happen when we expect the controller under test to modify state.